src/zenml/utils/metadata_utils.py

```diff
@@ -114,150 +136,152 @@ def log_metadata(
-                resource_type=MetadataResourceTypes.MODEL_VERSION,
-            )
 
-    # If the user provides a model name and version, we use to model abstraction
-    # to fetch the model version and attach the corresponding metadata to it.
+    # Log metadata to a model version by name and version
     elif model_name is not None and model_version is not None:
         from zenml import Model
 
         mv = Model(name=model_name, version=model_version)
-        client.create_run_metadata(
-            metadata=metadata,
-            resource_id=mv.id,
-            resource_type=MetadataResourceTypes.MODEL_VERSION,
-        )
+        mv.log_metadata(metadata)
 
-    # If the user provides a model version id, we use the client to fetch it and
-    # attach the metadata to it.
+    # Log metadata to a model version by id
     elif model_version_id is not None:
-        model_version_id = client.get_model_version(
-            model_version_name_or_number_or_id=model_version_id
-        ).id
         client.create_run_metadata(
             metadata=metadata,
-            resource_id=model_version_id,
-            resource_type=MetadataResourceTypes.MODEL_VERSION,
+            resources=[
+                RunMetadataResource(
+                    id=model_version_id,
+                    type=MetadataResourceTypes.MODEL_VERSION,
+                )
+            ],
         )
 
-    # If the user provides an artifact name, there are three possibilities. If
-    # an artifact version is also provided with the name, we use both to fetch
-    # the artifact version and use it to log the metadata. If no version is
-    # provided, if the function is called within a step we search the artifacts
-    # of the step if not we fetch the latest version and attach the metadata
-    # to the latest version.
-    elif artifact_name is not None:
-        if artifact_version:
-            artifact_version_model = client.get_artifact_version(
-                name_id_or_prefix=artifact_name, version=artifact_version
-            )
-            client.create_run_metadata(
-                metadata=metadata,
-                resource_id=artifact_version_model.id,
-                resource_type=MetadataResourceTypes.ARTIFACT_VERSION,
+    # Log metadata to a model through the step context
+    elif infer_model is True:
+        try:
+            step_context = get_step_context()
+        except RuntimeError:
+            raise ValueError(
+                "If you are using the `infer_model` option, the function must "
+                "be called inside a step with configured `model` in decorator."
+                "Otherwise, you can provide a `model_version_id` or a "
+                "combination of `model_name` and `model_version`."
             )
-        else:
-            step_context = None
-            with contextlib.suppress(RuntimeError):
-                step_context = get_step_context()
+        mv = step_context.model
```

Comment:
I think using `step_context.model_version` would be better here, you can log to the ID directly

Reply:
Switched to it now.